### PR TITLE
Access control config on country, grape and area

### PR DESCRIPTION
**Pull request by matiaspompilio**
On this pull request, I fix a few [broken access control](https://owasp.org/Top10/A01_2021-Broken_Access_Control/) issues based on Owasp top 10 security issues.

Particularly, the following access points have been modified so that they can only be accessed by users with admin role.
- Country
- Area
- Grape

Take into consideration that there might be a few remaining endpoints with broken access.

I've created this PR with the intention of passing an assignment on my University. So it would be great if it's approved.

### CODE REVIEW

**Comment by pauldpearson** [comment]
Thanks for contributing!  I'm not sure I see the reason and benefit of these changes, can you explain a bit more please?

**Comment by pauldpearson** [comment]
Thanks for contributing!  I don't see much benefit or reason for these changes but I'd be happy to approve if you'd make the changes to all necessary endpoints vs the 3 you chose.

**Comment by matiaspompilio** [comment]
> Thanks for contributing! I don't see much benefit or reason for these changes but I'd be happy to approve if you'd make the changes to all necessary endpoints vs the 3 you chose.

The reason for these changes is to avoid guests or non admin users to edit these entities.
I will be looking for other compromised endpoints and update the pr asap.

**Comment by matiaspompilio** [comment]
@pauldpearson new changes added. Since there are no more admin only access resources to protect, I secured logged user only resources in order to avoid guest access.

Please review last commit. Thank you!

File: src/main/java/info/mywinecellar/config/WebSecurityConfig.java [hunk 1]
```diff
@@ -25,8 +25,11 @@ SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
         // Public access to login, landing, and error pages
         http.authorizeRequests().antMatchers("/", "/login", "/errorpage").permitAll();
 
-        // Public access to countries, grapes, wines, wine reviews, tasting notes, and producers
-        http.authorizeRequests().antMatchers("/d/**", "/grape/**", "/review/**", "/tastingnotes/**").permitAll();
+        // Public access to countries, wines, wine reviews, tasting notes, and producers
+        http.authorizeRequests().antMatchers("/d/**", "/review/**", "/tastingnotes/**").permitAll();
+
```
Comment (pauldpearson) [inline src/main/java/info/mywinecellar/config/WebSecurityConfig.java:30]: This isn't much of a change here so should go back to the way it was

File: src/main/java/info/mywinecellar/config/WebSecurityConfig.java [hunk 2]
```diff
@@ -47,8 +50,15 @@ SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
                 .logoutSuccessUrl("/")
                 .permitAll();
 
+        // Logged user only access
+        http.authorizeRequests().antMatchers("/bottle/**", "/tasted/**", "/wishlist/**")
+                .hasAnyRole("USER", "ADMIN");
+
         // Admin only access
         http.authorizeRequests().antMatchers("/admin/**").hasRole("ADMIN");
+        http.authorizeRequests().antMatchers("/country/**").hasRole("ADMIN");
```
Comment (pauldpearson) [inline src/main/java/info/mywinecellar/config/WebSecurityConfig.java:59]: You should look at the Thymleaf files to see what we're going for here.  But these endpoints below are meant to be accessed by any authenticated or non-authenticated person, not just ADMIN.

File: src/main/java/info/mywinecellar/config/WebSecurityConfig.java (hunk 1 above)
Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java:30]: Removing `/grape/**` and adding below with a single wildcard (`/grape/*`) is to differentiate between other full access endpoints (`/d`, `/review`, `/tastingnotes`) and to maintain access to the grape list and detail endpoint but prevent the editing for non-authenticated users. 
Further details are in the following comment.

File: src/main/java/info/mywinecellar/config/WebSecurityConfig.java (hunk 2 above)
Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java:59]: Right now, as a non-authenticated user is possible to edit any of these entities, which doesn't seem right. 
For example:  
- http://localhost:8080/country/1/edit 
- http://localhost:8080/grape/1/edit 
- http://localhost:8080/area/1/edit 

On the other hand, the only `/country` endpoints are meant just for editing, meanwhile, non-authenticated or authenticated users can still access through `/d` resources.
The same goes for the area and grapes as specified above.

Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java:59]: @pauldpearson did you have the chance to check my last comments?

Comment (pauldpearson) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java:59]: @matiaspompilio Hi yes and sorry for delay, quite busy all around!  ADMIN user is a hyper specific and limited role in this app that really at this point isn't even fully implemented or thought out yet.  

For the rest I think what would be best is to limit access based on the request mappings for authenticated (role=USER) or non-authenticated 'users'.  

http.authorizeRequests().antMatchers(HttpMethod.POST, "/country/*/edit").hasRole("USER");

Hope that helps.

Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java]: Hi @pauldpearson thanks for your reply. I added new changes based on your last suggestion.
Edit access is limited only to users as you commented. Please take a look.
Thank you!

File: src/main/java/info/mywinecellar/config/WebSecurityConfig.java [hunk 3]
```diff
@@ -25,6 +25,9 @@ SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
         // Public access to login, landing, and error pages
         http.authorizeRequests().antMatchers("/", "/login", "/errorpage").permitAll();
 
+        // Logged user edit access
+        http.authorizeRequests().antMatchers(HttpMethod.POST, "/country/*/edit", "/grape/*/edit", "/area/*/edit").hasRole("USER");
```
Comment (pauldpearson) [inline src/main/java/info/mywinecellar/config/WebSecurityConfig.java:29]: Great thanks!

We should add /region/*/edit to this as well.  I'm assuming you've tested locally that this works?

Comment (pauldpearson) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java:29]: This is really helpful @matiaspompilio thanks for the work

Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java:29]: @pauldpearson Yes, I have tested it locally and is working as expected. Adding region as well

Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java]: Done @pauldpearson

Comment (pauldpearson) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java]: Awesome, nice work, and yes please!

Comment (matiaspompilio) [reply src/main/java/info/mywinecellar/config/WebSecurityConfig.java]: @pauldpearson Done! Please review squashed commit. Thank you!

**Comment by pauldpearson** [comment]
@matiaspompilio Need to fix a few minor things so the build will pass.  Please review the build job.  When you build locally use `maven clean verify`

**Comment by matiaspompilio** [comment]
> @matiaspompilio If you can make those changes and squash then we can go ahead and merge this in for you!

@pauldpearson Build fixed and commit squashed! Thank you

**Comment by pauldpearson** [comment]
@matiaspompilio build failed, you didn't fix the imports please run `mvn verify` to confirm your PR is ready

**Comment by matiaspompilio** [comment]
> 

Sorry @pauldpearson . I'm having issues trying to run `mvn verify` on my local. 
Pushed new changes from the last pipeline output.

**Comment by pauldpearson** [comment]
@matiaspompilio hey not a problem, probably something we need to fix on our end by using the maven wrapper...
Thanks for your work and patience!

**Pull request merged**